docker/api/exec_api.py

```diff
@@ -66,10 +64,18 @@ def exec_start(self, exec_id, detach=False, tty=False, stream=False,
             'Detach': detach
         }
 
+        headers = {} if detach else {
+            'Connection': 'Upgrade',
+            'Upgrade': 'tcp'
+        }
+
         res = self._post_json(
-            self._url('/exec/{0}/start', exec_id), data=data, stream=stream
+            self._url('/exec/{0}/start', exec_id),
+            headers=headers,
+            data=data,
+            stream=True
```

Comment:
Late answer, but yes, that's intentional. Without it, `_read_from_socket` cannot work properly. As far as the user is concerned, the output will still be presented according to their choice.